File: 3257-maximum-value-sum-by-placing-three-rooks-ii/3257-maximum-value-sum-by-placing-three-rooks-ii.py
class Solution:
    def maximumValueSum(self, board: List[List[int]]) -> int:
        n, m = len(board), len(board[0])
        l = []
        for i in range(n):
            cur = [(board[i][j], j) for j in range(m)]
            cur.sort(reverse=True)
            l.append(cur[:3])
        def process(x):
            best = []
            r = []
            for i in x:
                best += i
                best = sorted(best, reverse=True)
                t = []
                for e in best:
                    if any(j[1] == e[1] for j in t):
                        continue
                    t.append(e)
                    if len(t) == 3:
                        break
                best = t[:]
                r.append(best[:])
            return r
        pref, suff = process(l), process(l[::-1])[::-1]
        #print(pref)
        res = -10 ** 18
        for i in range(1, n - 1):
            for e, f in l[i]:
                for a, b in pref[i - 1]:
                    for c, d in suff[i + 1]:
                        if b != d and b != f and d != f:
                            res = max(res, a + c + e)
        return res

File: 3257-maximum-value-sum-by-placing-three-rooks-ii/3257-maximum-value-sum-by-placing-three-rooks-ii.java
class Solution {
    public long maximumValueSum(int[][] board) {
        int n = board.length;
        int m = board[0].length;
        long[] bs = new long[n*m];
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m; ++j) {
                long ta = board[i][j];
                ta=2000000000-ta;
                ta*=(n*m);
                ta+=i*m+j;
                bs[i*m+j]=ta;
            }
        }
        Arrays.sort(bs);
        long ans = -3000000000L;
        for (int i = 0; i < n*m; ++i) {
            int y1 = (int)(bs[i]%(n*m)/m);
            int x1 = (int)(bs[i]%(n*m)%m);
            long s1 = board[y1][x1];
            if (s1*3<=ans) {
                break;
            }
            for (int j = i+1; j < n*m; ++j) {
                int y2 = (int)(bs[j]%(n*m)/m);
                int x2 = (int)(bs[j]%(n*m)%m);
                long s2 = board[y2][x2];
                if (s1+s2*2<=ans) {
                    break;
                }
                if (y1==y2||x1==x2) {
                    continue;
                }
                for (int k = j+1; k < n*m; ++k) {
                    int y3 = (int)(bs[k]%(n*m)/m);
                    int x3 = (int)(bs[k]%(n*m)%m);
                    long s3 = board[y3][x3];
                    if (s1+s2+s3<=ans) {
                        break;
                    }
                    if (y1==y3||y2==y3||x1==x3||x2==x3) {
                        continue;
                    }
                    ans = s1+s2+s3;
                }
            }
        }
        return ans;
    }
}